import os, glob

configfile: "../config/config.json"

include: "rules/common.smk"
include: "rules/downloads.smk"
include: "rules/sno_literature_processing.smk"
include: "rules/sno_tgirt_seq_processing.smk"
include: "rules/sno_families.smk"
include: "rules/negative_examples.smk"
include: "rules/tune_train_test_sets.smk"
include: "rules/one_hot_encoding_and_normalization.smk"


rule all:
    input:
        normalized_test = 'data/references/positives_and_negatives/initial/initial_encoded_scaled_test_set.tsv'

rule all_downloads:
    input:
        blat_installation_file = "data/references/blat/blat_test.txt",
        genome = expand('data/references/genome_fa/{species}_genome.fa', **config),
        bed = expand('data/references/rnacentral/{species}.bed', 
                species=[sp for sp in config['species']+config['species_tgirt'] 
                        if sp not in ['ostreococcus_tauri', 'schizosaccharomyces_pombe']]),
        gallus_gallus_gff = "data/references/rnacentral/gallus_gallus_bGalGal1_GRCg7b.gff3",
        gtf = expand('data/references/gtf/{species}.gtf', species=[sp for sp in 
                        config['species']+config['species_tgirt'] if sp not in 
                        ['ostreococcus_tauri', 'schizosaccharomyces_pombe']]),
        rfam_clans = "data/references/rfam/fam_clans_families.tsv"




